tests/unit/test_charm.py

```diff
@@ -161,6 +184,36 @@ def test_layer_updated_without_any_endpoint_info(harness: Harness) -> None:
     assert harness.charm._login_ui_layer.to_dict() == expected_layer
 
 
+def test_layer_updated_with_tracing_endpoint_info(harness: Harness) -> None:
+    """Test Pebble Layer when relation data is in place."""
+    harness.set_leader(True)
+    harness.set_can_connect(CONTAINER_NAME, True)
+    harness.charm.on.login_ui_pebble_ready.emit(CONTAINER_NAME)
+    tracing_relation_id = setup_tracing_relation(harness)
+
+    host = harness.get_relation_data(tracing_relation_id, "tempo-k8s")["host"]
+    http = filter(
+        lambda x: x["protocol"] == "otlp_http",
+        harness.get_relation_data(tracing_relation_id, "tempo-k8s")["ingesters"],
+    )
+    grpc = filter(
+        lambda x: x["protocol"] == "otlp_grpc",
+        harness.get_relation_data(tracing_relation_id, "tempo-k8s")["ingesters"],
+    )
+    assert (
```

Comment:
I think the assertion here probably will fail because the `filter` function returns a `filter` object. We cannot do a comparison between a `filter` object and a `str`.

If we need to filter the data in the `databag`, we could do something like:
```python
actual = (data['protocol'] for data in databag if data['protocol'] == 'otlp_grpc')
assert expected == next(actual)
```

Reply:
```
    host = json.loads(relation["host"])
    ingesters = json.loads(relation["ingesters"])
    http = [i["port"] for i in ingesters if i["protocol"] == "otlp_http"][0]
    grpc = [i["port"] for i in ingesters if i["protocol"] == "otlp_grpc"][0]
```

this doesn't seem right, i would expect the databag to be deserialized at this point, am i wrong?

anyway all works now

Reply:
Probably needs to look at the harness API to find out how it does the translations. I saw you used the method provided in the tracing lib which is a nice and neat API to do the job :thumbsup: